thunder/torch/__init__.py

```diff
@@ -1812,6 +1812,25 @@ def leaky_relu(a: TensorProxy, /, negative_slope: float = 0.01, inplace: bool =
 _inplace_to_out_of_place[leaky_relu] = leaky_relu, 2
 
 
+@torchsymbol(torch.nn.functional.logsigmoid, is_method=False)
+def logsigmoid(a: TensorProxy, /) -> TensorLike:
+    return where(a > 0, -log1p(exp(-a)), a - log1p(exp(a)))
+
+
+_inplace_to_out_of_place[logsigmoid] = logsigmoid, -1
+
+
+# @torchsymbol("log_sigmoid_backward", id="log_sigmoid_backward")
```

Comment:
Why, oh why, oh why does test_grad fail on the cpu and only the cpu when this line is uncommented?

Reply:
I'm not sure what the torchsymbol addition actually does in this case, but as an fyi here are the logsigmoid backward implementations in PyTorch:

https://github.com/pytorch/pytorch/blob/0ddb33ba2299542f9558d949aa482a9ca30ceb30/aten/src/ATen/native/cuda/ActivationLogSigmoidKernel.cu#L43

https://github.com/pytorch/pytorch/blob/0ddb33ba2299542f9558d949aa482a9ca30ceb30/aten/src/ATen/native/cpu/Activation.cpp#L94

Reply:
Yes, I was looking at those, and also:

https://github.com/pytorch/pytorch/blob/fd65bd755d30f69783f82c75f9f341c1bbd4614a/torch/_decomp/decompositions.py#L332

That's why there's this extra unused parameter in the signature (for `buffer` which is also unused in PyTorch), and why I register the backward op with 
```
log_sigmoid_backward = _register_torch_operation(
    "torch.ops.aten.log_sigmoid_backward", like=ltorch.log_sigmoid_backward
)
```

I also don't understand what the annotation does here, but it looks like what was done for `log_softmax_backward` [here](https://github.com/Lightning-AI/lightning-thunder/blob/f89cecac3b8a7e1e67913b1f0b53fa4c3b8322e8/thunder/torch/__init__.py#L4963), which has backward implementations in PyTorch too:

https://github.com/pytorch/pytorch/blob/fd65bd755d30f69783f82c75f9f341c1bbd4614a/torch/_decomp/decompositions.py#L854

I was also looking at `_nll_loss_backward_impl` and wondering about [this comment](https://github.com/Lightning-AI/lightning-thunder/blob/f89cecac3b8a7e1e67913b1f0b53fa4c3b8322e8/thunder/executors/torchex.py#L1612).

Reply:
Aha. I'm guessing the `torchsymbol` decorator registers the operator as a symbol so it can be consumed directly by an executor later. That comment probably says that the `impl` would ideally not be defined in the torch executor but available in the torch language definition, like where the log sigmoid is in this PR. 

Is `torch.ops.aten.log_sigmoid_backward` actually getting called as expected if you look at the traces? Do you think the forward->backward computations are actually different (when trying them) or do you think the test has a problem?

Reply:
Running the following program:
```
import thunder
import torch

def f(x):
    return torch.nn.functional.logsigmoid(x)

x = torch.rand((1024,1024,4), device='cpu', requires_grad=True)
jf = thunder.jit(f)
jf(x)
print(thunder.last_backward_traces(jf)[-1])
```
 with the annotation produces:
```
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, _, = saved_for_backward
  clear_mutable_collection(saved_for_backward)
  del saved_for_backward
  t18, = cotangents
  clear_mutable_collection(cotangents)
  del cotangents
  x, = C0
  clear_mutable_collection(C0)
  del C0
  t27 = torch.torch.ops.aten.log_sigmoid_backward(t18, x, x)  # t27: "cpu f32[1024, 1024, 4]"
    # t27 = ltorch.log_sigmoid_backward(t18, x, x)  # t27: "cpu f32[1024, 1024, 4]"
      # t37 = ltorch.abs(x)  # t37: "cpu f32[1024, 1024, 4]"
        # t37 = prims.abs(x)  # t37: "cpu f32[1024, 1024, 4]"
      # t38 = ltorch.neg(t37)  # t38: "cpu f32[1024, 1024, 4]"
        # t38 = prims.neg(t37)  # t38: "cpu f32[1024, 1024, 4]"
      # t39 = ltorch.exp(t38)  # t39: "cpu f32[1024, 1024, 4]"
        # t39 = prims.exp(t38)  # t39: "cpu f32[1024, 1024, 4]"
      # t40 = ltorch.add(1, t39, alpha=1)  # t40: "cpu f32[1024, 1024, 4]"
        # t40 = prims.add(1.0, t39)  # t40: "cpu f32[1024, 1024, 4]"
      # t41 = ltorch.true_divide(t39, t40)  # t41: "cpu f32[1024, 1024, 4]"
        # t41 = prims.div(t39, t40)  # t41: "cpu f32[1024, 1024, 4]"
      # t42 = ltorch.gt(x, 0)  # t42: "cpu b8[1024, 1024, 4]"
        # t42 = prims.gt(x, 0.0)  # t42: "cpu b8[1024, 1024, 4]"
      # t43 = ltorch.sub(1, t41, alpha=1)  # t43: "cpu f32[1024, 1024, 4]"
        # t43 = prims.sub(1.0, t41)  # t43: "cpu f32[1024, 1024, 4]"
      # t44 = ltorch.where(t42, t41, t43)  # t44: "cpu f32[1024, 1024, 4]"
        # t44 = prims.where(t42, t41, t43)  # t44: "cpu f32[1024, 1024, 4]"
      # t27 = ltorch.mul(t18, t44)  # t27: "cpu f32[1024, 1024, 4]"
        # t27 = prims.mul(t18, t44)  # t27: "cpu f32[1024, 1024, 4]"
  del t18, x
  return (t27,)]
```

While the results of `test_grad` are
```
>       comp(J_u_v, u_J_star_v)
E       AssertionError: Scalars are not close!
E
E       Expected 0.0716214267429005 but got 0.22672571966112479.
E       Absolute difference: 0.15510429291822428 (up to 1e-07 allowed)
E       Relative difference: 2.1656130011903043 (up to 1e-07 allowed)

...

FAILED thunder/tests/test_grad.py::test_vjp_correctness_logsigmoid_torch_cpu_thunder.dtypes.float64 - AssertionError: Scalars are not close!
```

Without the annotation:
```
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, _, = saved_for_backward
  clear_mutable_collection(saved_for_backward)
  del saved_for_backward
  t18, = cotangents
  clear_mutable_collection(cotangents)
  del cotangents
  x, = C0
  clear_mutable_collection(C0)
  del C0
  t19 = torch.abs(x)  # t19: "cpu f32[1024, 1024, 4]"
    # t19 = ltorch.abs(x)  # t19: "cpu f32[1024, 1024, 4]"
      # t19 = prims.abs(x)  # t19: "cpu f32[1024, 1024, 4]"
  t24 = torch.gt(x, 0)  # t24: "cpu b8[1024, 1024, 4]"
    # t24 = ltorch.gt(x, 0)  # t24: "cpu b8[1024, 1024, 4]"
      # t24 = prims.gt(x, 0.0)  # t24: "cpu b8[1024, 1024, 4]"
  del x
  t20 = torch.neg(t19)  # t20: "cpu f32[1024, 1024, 4]"
    # t20 = ltorch.neg(t19)  # t20: "cpu f32[1024, 1024, 4]"
      # t20 = prims.neg(t19)  # t20: "cpu f32[1024, 1024, 4]"
  del t19
  t21 = torch.exp(t20)  # t21: "cpu f32[1024, 1024, 4]"
    # t21 = ltorch.exp(t20)  # t21: "cpu f32[1024, 1024, 4]"
      # t21 = prims.exp(t20)  # t21: "cpu f32[1024, 1024, 4]"
  del t20
  t22 = torch.add(1, t21, alpha=1)  # t22: "cpu f32[1024, 1024, 4]"
    # t22 = ltorch.add(1, t21, alpha=1)  # t22: "cpu f32[1024, 1024, 4]"
      # t22 = prims.add(1.0, t21)  # t22: "cpu f32[1024, 1024, 4]"
  t23 = torch.true_divide(t21, t22)  # t23: "cpu f32[1024, 1024, 4]"
    # t23 = ltorch.true_divide(t21, t22)  # t23: "cpu f32[1024, 1024, 4]"
      # t23 = prims.div(t21, t22)  # t23: "cpu f32[1024, 1024, 4]"
  del t21, t22
  t25 = torch.sub(1, t23, alpha=1)  # t25: "cpu f32[1024, 1024, 4]"
    # t25 = ltorch.sub(1, t23, alpha=1)  # t25: "cpu f32[1024, 1024, 4]"
      # t25 = prims.sub(1.0, t23)  # t25: "cpu f32[1024, 1024, 4]"
  t26 = torch.where(t24, t23, t25)  # t26: "cpu f32[1024, 1024, 4]"
    # t26 = ltorch.where(t24, t23, t25)  # t26: "cpu f32[1024, 1024, 4]"
      # t26 = prims.where(t24, t23, t25)  # t26: "cpu f32[1024, 1024, 4]"
  del t24, t23, t25
  t27 = torch.mul(t18, t26)  # t27: "cpu f32[1024, 1024, 4]"
    # t27 = ltorch.mul(t18, t26)  # t27: "cpu f32[1024, 1024, 4]"
      # t27 = prims.mul(t18, t26)  # t27: "cpu f32[1024, 1024, 4]"
  del t18, t26
  return (t27,)]
```

The issue is not a misapplication of `fdm`.   It produces exactly the results I expect.  Testing with a random 1 element tensor shows that the vjp of logsigmoid on the cpu is off.  Meaning it's not some cumulative discrepancy for large tensors, nor an issue with a singularity.  

If I use the one line implementation (without the `abs`) of `log_sigmoid_backward`, traces of the annotated and unannotated version change as one would expect, with `test_grad` still failing for the decorated version.

Reply:
That's so interesting! So it looks like the logsigmoid foward and backward are being called as expected, and we 

So if you just forward+backward without going through test vjp on some examples, how do the grads compare between the two? Are there any obvious visual differences for small values? (Also, are the forward traces what you'd expect?) What if you explicitly test calling logsigmoid + logsigmoid_backward in eager PyTorch? How does that compare to what eager is actually doing?

Reply:
One idea that shouldn't make a difference, by the way, would be to write the fwd+bwd as a single function, like this one: https://github.com/Lightning-AI/lightning-thunder/blob/96486e81cf8b5c5a90faa3b2a529669d74d59658/thunder/core/transforms.py#L682

```
from thunder.torch import logsigmoid, log_sigmoid_backward
  fwd = logsigmoid(a)
  g = get_grad(fwd)
  ga = log_sigmoid_backward(g, a, _)

    g = get_grad(fwd)
    g_converted = prims.convert_element_type(g, dtypes.to_dtype(a))
    put_grad(a, g_converted)

    return fwd
```